multi.c

```diff
@@ -0,0 +1,145 @@
+#include "multi.h"
+
+#include "brute.h"
+#include "queue.h"
+#include "single.h"
+#include <assert.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+void *
+mt_password_check (void *context)
+{
+  mt_context_t *mt_context = (mt_context_t *)context;
+  task_t task;
+  st_context_t st_context = {
+    .hash = mt_context->config->hash,
+    .data = { .initialized = 0 },
+  };
+
+  while (true)
+    {
+      /* don't understand why we need status check here */
+      if (queue_pop (&mt_context->queue, &task) == S_FAILURE)
+        {
+          return NULL;
+        }
+      if (st_password_check (&task, &st_context))
+        {
+          /* This is printing but loop doesn't break */
+          /* printf ("Password found: %s\n", task.password); */
+          memcpy (mt_context->password, task.password, sizeof (task.password));
+        }
+
+      if (pthread_mutex_lock (&mt_context->mutex) != 0)
+        {
+          return NULL;
+        }
+      --mt_context->passwords_remaining;
+      /* printf("-- %d\n", mt_context->passwords_remaining); */
+      if (mt_context->passwords_remaining == 0)
+        {
+          if (pthread_cond_signal (&mt_context->cond_sem) != 0)
+            {
+              return NULL;
+            }
+        }
+      if (pthread_mutex_unlock (&mt_context->mutex) != 0)
+        {
+          return NULL;
+        }
+    }
+  return NULL;
+}
+
+bool
+queue_push_wrapper (task_t *task, void *context)
+{
+  mt_context_t *mt_context = (mt_context_t *)context;
+  if (queue_push (&mt_context->queue, task) == S_FAILURE)
+    {
+      return false;
+    }
+
+  if (pthread_mutex_lock (&mt_context->mutex) != 0)
+    {
+      return false;
+    }
+  ++mt_context->passwords_remaining;
+  if (pthread_mutex_unlock (&mt_context->mutex) != 0)
+    {
+      return false;
+    }
+  /* printf("++ %d\n", mt_context->passwords_remaining); */
+
+  return mt_context->password[0] != 0;
+}
+
+// TODO: Find memory leak and fix it
```

Comment:
Memory allocation happens here only in pthread_create. Deallocation should happen in pthread_join.
Any return from the function in between create/join will cause memory leak.
There is no evident memory leak in a happy path scenario. Leakage is possible only in case of errors. The proper way to handle those scenarios is to use pthread_cleanup_push/pthread_cleanup_pop.

Use valgrind to trace allocations and find particular leakage.

Reply:
> Memory allocation happens here only in pthread_create. Deallocation should happen in pthread_join. Any return from the function in between create/join will cause memory leak. There is no evident memory leak in a happy path scenario. Leakage is possible only in case of errors. The proper way to handle those scenarios is to use pthread_cleanup_push/pthread_cleanup_pop.
> 
> Use valgrind to trace allocations and find particular leakage.

I found the possible memory leak by valgrind in some of the previous commits, but got rid of it later by fixing some of the bugs I guess, so now there is no possible memory leaks as valgrind says.